gcc10 multiple symbol definitions (-fno-common)

<details>
<summary>CRAN additonal issue gcc10</summary>
* installing *source* package â€˜data.tableâ€™ ...
** package â€˜data.tableâ€™ successfully unpacked and MD5 sums checked
** using staged installation
zlib 1.2.11 is available ok
** libs
... snip ...
/usr/local/gcc10/bin/gcc -I"/data/gannet/ripley/R/R-gcc10/include" -DNDEBUG   -I/usr/local/include  -fopenmp -fpic  -g -O2 -Wall -pedantic -mtune=native -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong -fstack-clash-protection -fcf-protection  -c wrappers.c -o wrappers.o
/usr/local/gcc10/bin/gcc -shared -L/usr/local/gcc10/lib64 -L/usr/local/lib64 -o data.table.so assign.o between.o bmerge.o chmatch.o cj.o coalesce.o dogroups.o fastmean.o fcast.o fifelse.o fmelt.o forder.o frank.o fread.o freadR.o froll.o frollR.o frolladaptive.o fsort.o fwrite.o fwriteR.o gsumm.o ijoin.o init.o inrange.o nafill.o nqrecreateindices.o openmp-utils.o quickselect.o rbindlist.o reorder.o shift.o subset.o transpose.o types.o uniqlist.o utils.o vecseq.o wrappers.o -fopenmp -lz
/usr/bin/ld: between.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:67: multiple definition of `char_integer64'; assign.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:67: first defined here
/usr/bin/ld: between.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:106: multiple definition of `SelfRefSymbol'; assign.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:106: first defined here
multiple definition of `sym_verbose'; assign.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:88: first defined here
/usr/bin/ld: frank.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:87: multiple definition of `sym_colClassesAs'; assign.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:87: first defined here
/usr/bin/ld: frank.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:86: multiple definition of `sym_maxgrpn'; assign.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:86: first defined here

... hundreds of similar, snipped ...

/usr/bin/ld: wrappers.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:68: multiple definition of `char_ITime'; assign.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:68: first defined here
/usr/bin/ld: wrappers.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:67: multiple definition of `char_integer64'; assign.o:/data/gannet/ripley/R/packages/tests-gcc10/data.table/src/data.table.h:67: first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [/data/gannet/ripley/R/R-gcc10/share/make/shlib.mk:6: data.table.so] Error 1
make[2]: Target 'all' not remade because of errors.
make[2]: Leaving directory '/data/gannet/ripley/R/packages/tests-gcc10/data.table/src'
ERROR: compilation failed for package â€˜data.tableâ€™
* removing â€˜/data/gannet/ripley/R/packages/tests-gcc10/data.table.Rcheck/data.tableâ€™
</details>

This concerns packages

BNSP BPEC C50 CompRandFld Cubist HWxtest MigClim PBSddesolve PET
PwrGSD RDieHarder Rmpfr TTR VBLPCM actuar aucm bibtex bitrugs carfima
cec2005benchmark cleancall coxme cubature curl data.table deSolve
deTestSet dti eha epanet2toolkit fansi gensvm geometry gpg hdf5r hergm
keypress landscapemetrics latentnet lfe memisc mmap msm networkDynamic
pbdMPI pgmm phyclust phylocomr pomdp pop.wolf population randtoolbox
rootSolve samon smoof soobench topicmodels vctrs vetr weightedScores xts

which fail to install with gcc trunk which is due to be released as gcc
10.0.x early in 2020.  It defaults to -fno-common, which stricter on
some multiply-defined symbols -- see §1.6.4.1 of 'Writing R Extensions'
in current R-devel.

See https://www.stats.ox.ac.uk/pub/bdr/gcc10/ for the logs, which will
be linked from the CRAN results page under 'Additional issues'.

We expect the issues to also be seen when compiling with earlier
versions of gcc using the (non-default) flag -fno-common: this has been
confirmed for gcc 9 on Linux, and also for clang on macOS.

Please correct as soon as possible and before Jan 10 to safely retain on
CRAN.  (This is after the CRAN shutdown from Dec 20 to Jan 6.)

--
Brian D. Ripley
Emeritus Professor of Applied Statistics, University of Oxford

--- a/.dev/cc.R
+++ b/.dev/cc.R
@@ -62,7 +62,7 @@ cc = function(test=TRUE, clean=FALSE, debug=FALSE, omp=!debug, cc_dir, path=Sys.
   if (debug) {
     ret = system(sprintf("MAKEFLAGS='-j CC=%s PKG_CFLAGS=-f%sopenmp CFLAGS=-std=c99\\ -O0\\ -ggdb\\ -pedantic' R CMD SHLIB -d -o data.table.so *.c", CC, OMP))
   } else {
-    ret = system(sprintf("MAKEFLAGS='-j CC=%s CFLAGS=-f%sopenmp\\ -std=c99\\ -O3\\ -pipe\\ -Wall\\ -pedantic' R CMD SHLIB -o data.table.so *.c", CC, OMP))
+    ret = system(sprintf("MAKEFLAGS='-j CC=%s CFLAGS=-f%sopenmp\\ -std=c99\\ -O3\\ -pipe\\ -Wall\\ -pedantic\\ -fno-common' R CMD SHLIB -o data.table.so *.c", CC, OMP))
     # TODO add -Wextra too?
   }
   if (ret) return()

--- a/src/data.table.h
+++ b/src/data.table.h
@@ -65,36 +65,40 @@ typedef R_xlen_t RLEN;
 #endif
 
 // init.c
-SEXP char_integer64;
-SEXP char_ITime;
-SEXP char_IDate;
-SEXP char_Date;
-SEXP char_POSIXct;
-SEXP char_nanotime;
-SEXP char_lens;
-SEXP char_indices;
-SEXP char_allLen1;
-SEXP char_allGrp1;
-SEXP char_factor;
-SEXP char_ordered;
-SEXP char_datatable;
-SEXP char_dataframe;
-SEXP char_NULL;
-SEXP sym_sorted;
-SEXP sym_index;
-SEXP sym_BY;
-SEXP sym_starts, char_starts;
-SEXP sym_maxgrpn;
-SEXP sym_colClassesAs;
-SEXP sym_verbose;
-SEXP sym_inherits;
-SEXP sym_datatable_locked;
+extern SEXP char_integer64;
+extern SEXP char_ITime;
+extern SEXP char_IDate;
+extern SEXP char_Date;
+extern SEXP char_POSIXct;
+extern SEXP char_nanotime;
+extern SEXP char_lens;
+extern SEXP char_indices;
+extern SEXP char_allLen1;
+extern SEXP char_allGrp1;
+extern SEXP char_factor;
+extern SEXP char_ordered;
+extern SEXP char_datatable;
+extern SEXP char_dataframe;
+extern SEXP char_NULL;
+extern SEXP sym_sorted;
+extern SEXP sym_index;
+extern SEXP sym_BY;
+extern SEXP sym_starts, char_starts;
+extern SEXP sym_maxgrpn;
+extern SEXP sym_colClassesAs;
+extern SEXP sym_verbose;
+extern SEXP SelfRefSymbol;
+extern SEXP sym_inherits;
+extern SEXP sym_datatable_locked;
+extern double NA_INT64_D;
+extern long long NA_INT64_LL;
+extern Rcomplex NA_CPLX;  // initialized in init.c; see there for comments
+extern size_t sizes[100];  // max appears to be FUNSXP = 99, see Rinternals.h
+extern size_t typeorder[100];
+
 long long DtoLL(double x);
 double LLtoD(long long x);
 bool GetVerbose();
-double NA_INT64_D;
-long long NA_INT64_LL;
-Rcomplex NA_CPLX;  // initialized in init.c; see there for comments
 
 // cj.c
 SEXP cj(SEXP base_list);
@@ -102,9 +106,6 @@ SEXP cj(SEXP base_list);
 // dogroups.c
 SEXP keepattr(SEXP to, SEXP from);
 SEXP growVector(SEXP x, R_len_t newlen);
-size_t sizes[100];  // max appears to be FUNSXP = 99, see Rinternals.h
-size_t typeorder[100];
-SEXP SelfRefSymbol;
 
 // assign.c
 SEXP allocNAVector(SEXPTYPE type, R_len_t n);

--- a/src/init.c
+++ b/src/init.c
@@ -3,6 +3,39 @@
 #include <R_ext/Rdynload.h>
 #include <R_ext/Visibility.h>
 
+// global constants extern in data.table.h for gcc10 -fno-common; #4091
+// these are written to once here on initialization, but because of that write they can't be declared const
+SEXP char_integer64;
+SEXP char_ITime;
+SEXP char_IDate;
+SEXP char_Date;
+SEXP char_POSIXct;
+SEXP char_nanotime;
+SEXP char_lens;
+SEXP char_indices;
+SEXP char_allLen1;
+SEXP char_allGrp1;
+SEXP char_factor;
+SEXP char_ordered;
+SEXP char_datatable;
+SEXP char_dataframe;
+SEXP char_NULL;
+SEXP sym_sorted;
+SEXP sym_index;
+SEXP sym_BY;
+SEXP sym_starts, char_starts;
+SEXP sym_maxgrpn;
+SEXP sym_colClassesAs;
+SEXP sym_verbose;
+SEXP SelfRefSymbol;
+SEXP sym_inherits;
+SEXP sym_datatable_locked;
+double NA_INT64_D;
+long long NA_INT64_LL;
+Rcomplex NA_CPLX;
+size_t sizes[100];
+size_t typeorder[100];
+
 // .Calls
 SEXP setattrib();
 SEXP bmerge();

